Editorial review: Document meta text-scale and env(preferred-text-scale) (#43766)
mdn/content: files/en-us/web/css/reference/values/env/index.md

      - : The dimensions of a visible `titlebar-area-*` area. These variables are available when using the `window-controls-overlay` [`display_override`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/display_override) manifest field. The variables' values can be used to ensure content doesn't overlap window control buttons (that is, minimize, maximize, and close) with progressive web apps (PWA) installed on desktop devices.
    - `keyboard-inset-top`, `keyboard-inset-right`, `keyboard-inset-bottom`, `keyboard-inset-left`, `keyboard-inset-width`, `keyboard-inset-height`
      - : The insets from the edge of the viewport and dimensions of the device's on-screen virtual keyboard. Defined in the {{domxref("VirtualKeyboard API", "VirtualKeyboard API", "", "nocode")}}.
    - `preferred-text-scale`

>       - : A number indicating the user's preferred font scaling factor, as set by browser or OS-level preferences. This can be used to size website text or other UI features proportionally to browser or OS-set font sizes.

Reviewer: ```suggestion
      - : A number indicating the user's preferred font scaling factor, as set by browser or OS-level preferences. This can be used to size content proportionally to browser or OS-set font sizes.
```

Reviewer: changed to make it the briefest explanation possible here, and the removed content is iterated below.

      - : The dimensions of a visible `titlebar-area-*` area. These variables are available when using the `window-controls-overlay` [`display_override`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/display_override) manifest field. The variables' values can be used to ensure content doesn't overlap window control buttons (that is, minimize, maximize, and close) with progressive web apps (PWA) installed on desktop devices.
    - `keyboard-inset-top`, `keyboard-inset-right`, `keyboard-inset-bottom`, `keyboard-inset-left`, `keyboard-inset-width`, `keyboard-inset-height`
      - : The insets from the edge of the viewport and dimensions of the device's on-screen virtual keyboard. Defined in the {{domxref("VirtualKeyboard API", "VirtualKeyboard API", "", "nocode")}}.
    - `preferred-text-scale`

>       - : A number indicating the user's preferred font scaling factor, as set by browser or OS-level preferences. This can be used to size content proportionally to browser or OS-set font sizes.

Reviewer: original could be confusing for some who may think you set a number.  how about:

```suggestion
      - : The user's preferred font scaling factor, a number set in browser or OS-level preferences. This can be used to size content proportionally to browser or OS-set font sizes.
```

The `viewport-segment-*` variable names can be used to set your containers to fit neatly into the available segments of a multi-viewport-segment device such as a hinged or foldable device. The integers following the `viewport-segment-*` name indicate which segment of the multiple segments the environment variable is referencing.

The `preferred-text-scale` variable can be used to size website text or other UI features proportionally to browser or OS-set font sizes. For example, you could set your body font size to be in proportion to a user-defined text scale like so:

```css
body {

>   font-size: calc(100% * env(preferred-text-scale));

Reviewer: This is multiplying by 1, which appears to be the same as `font-size: env(preferred-text-scale);` so this needs to be explained.
 did you mean `font-size-adjust`?

Author: It is only multiplying by `1` when the OS font size is set to "medium" (Chrome desktop), or whatever the equivalent is in other OSes. I have seen examples using `font-size`, and it does work, but it might be good to get some input from the engineers about recommended usage.

Reviewer 2: Our recommended usage is that sites include `<meta name="text-scale" content="scale" />`, then
1. Don't set the root font size
2. Use only `(r)em` or the font size [keywords](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/font-size#syntax) ('small', 'x-large', etc) to size content
3. Don't use `env(p...)`

(I referred to https://www.joshtumath.uk/posts/2026-01-27-try-text-scaling-support-in-chrome-canary/ to write that out.)

If a site follows those rules, we think the browser will size text properly according to the user's wishes across both mobile and desktop.

However, for some sites, especially complex or older sites, "Use only `(r)em` to size content" is not possible. In that case, they should use `env(preferred-text-scale)` inside calc expressions to inflate the parts of the page they know the user will want to most easily read. e.g. reddit uses `text-size-adjust:calc(100% * env(preferred-text-scale))` on comment text but apparently not the number of votes.

<img width="399" height="549" alt="Image" src="https://github.com/user-attachments/assets/5f485e69-1634-441e-b38c-1ba9d03ad9e3" />

The reddit screenshot is taken with an emulated 1.5x font scale set via chrome devtools.

<img width="455" height="482" alt="Image" src="https://github.com/user-attachments/assets/06b644eb-fc1e-466c-b18b-21f5d6c7431c" />

Reviewer 2: @joshtumath JFYI

Author: Great, thanks, @davidsgrogan. I've added the recommended usage details to the `<meta name="text-scale">` page, and more strongly stated that you shouldn't use the `env()` variable.

}
```

Effects on supported browsers:

- Chrome for Android

>   - : The body font size will be set in proportion to the global OS "Font size" setting.

Reviewer: ```suggestion
For example, on Chrome for Android, the body font size will be set in proportion to the global OS "Font size" setting.
```
future proof

> [!NOTE]

> > You can also set page sizes to be proportional to browser or OS-level font settings by including the [`<meta name="text-scale" content="scale">`](/en-US/docs/Web/HTML/Reference/Elements/meta/name/text-scale) tag inside the page's `<head>`. You are advised to use the `<meta>` tag rather than the `env()` variable, provided it is supported in your target browser set and you are able to edit the HTML. The `<meta>` tag also has some helpful effects on desktop browsers.

Reviewer: Suggest making this more reference-ish ( less guide-ish):

 ```suggestion
> Page sizes can also be set to be proportional to browser or OS-level font size by including [`<meta name="text-scale" content="scale">`](/en-US/docs/Web/HTML/Reference/Elements/meta/name/text-scale) inside the document `<head>`. The `<meta>` tag should be used in favor of the `env(preferred-text-scale)` when possible, as the `<meta>` tag has some helpful effects on desktop browsers.
```

That said, i have some questions:

"has some helpful effects on  desktop browsers" is really vague. What is(are) the benefit(s)? 
If it's just better supported, use "legacy" language. (If this page isn't updated for 10 years, what are we telling future devs)

Reviewer: FOund it:

"The [preferred-text-scale](https://drafts.csswg.org/css-env-1/#valdef-env-preferred-text-scale) environment variable returns the user’s OS-level font factor on mobile devices, but 1 on desktop devices. [text scale factor](https://drafts.csswg.org/css-fonts-5/#text-scale-factor) incorporates UA-level font preferences but OS-level font preferences are ignored." - this is why meta is better...it sets it for all.

Author: I've committed your suggestion, as I think it sounds better. For the moment, I've updated the note to:

> [!NOTE]
> Page sizes can also be set to be proportional to browser or OS-level font size by including [`<meta name="text-scale" content="scale">`](/en-US/docs/Web/HTML/Reference/Elements/meta/name/text-scale) inside the document `<head>`. The `<meta>` tag should be used in favor of the `env(preferred-text-scale)` when possible, as the `<meta>` tag is simpler to use and is supported across a wider range of platforms.

The thing is, I'm not sure this quote you've found from the spec is quite how it works in current implementations. Both of them have the effect of respecting OS-level font settings on Android, for example.

Again, something else that it would be good to get engineer input on.

Author: @davidsgrogan, can I get your input on this thread too? Thanks!

>

> > You are also advised not to use both — this results in the text scaling unexpectedly, with small font sizes being smaller and large font sizes being larger.

Reviewer: My understanding, which could be totally wrong, is  `env(preferred-text-scale)` returns the user’s OS-level font preferences, and the text scale factor uses both the UA and OS font, so including both would just be saying "scale" twice, so there would be no double effect... or is it applied twice?

if we do include it, how about:
```suggestion
> Avoid including both these features as this may result in redundant text scaling, with small font sizes being smaller and large font sizes being larger.
```

Reviewer: didn't realize this was tech review and not editorial....

Author: I think your change sounds reasonable. I've asked the engineers what exactly is going on when you set both, but it does seem like redundancy.

Reviewer 2: It would be applied twice. (I think this is the same case we talked about over e-mail?)

Author: Cool, thanks for the clarification. I've updated this bit to say that the text scaling would be applied twice.

The `viewport-segment-*` variable names can be used to set your containers to fit neatly into the available segments of a multi-viewport-segment device such as a hinged or foldable device. The integers following the `viewport-segment-*` name indicate which segment of the multiple segments the environment variable is referencing.

> The `preferred-text-scale` variable can be used to size website text or other UI features proportionally to browser or OS-set font sizes, such as Chrome for Android's global OS "Font size" setting. For example, you could set the body font size to be a percentage based on the user-defined text scale:

Reviewer: thanks. that makes this make sense.

```css
body {
  font-size: calc(100% * env(preferred-text-scale));
}
```

> Page sizes can also be set to be proportional to browser or OS-level font size by including [`<meta name="text-scale" content="scale">`](/en-US/docs/Web/HTML/Reference/Elements/meta/name/text-scale) inside the document `<head>`. The `<meta>` tag should be used in favor of the `env(preferred-text-scale)` when possible, as the `<meta>` tag is simpler to use and is supported across a wider range of platforms.

Reviewer: ```suggestion
Sizes can also be set to be proportional to browser or OS-level font size by including [`<meta name="text-scale" content="scale">`](/en-US/docs/Web/HTML/Reference/Elements/meta/name/text-scale) inside the document `<head>`. The `<meta>` tag should be used in favor of the `env(preferred-text-scale)` when possible, as the `<meta>` tag is supported across a wider range of platforms and is also simpler to use.
```
It's not the "page size", but "sizes" isn't good either. maybe "page tex-size scaling"?
re-arranged the order of the last sentence b/c the support is the reason, the simpler use is an added bonus.

Author: I think "Sizes" is OK. We are basically talking about font-proportionate dimensions; something like "page text-size scaling" is too complex and will confuse people.

> [!WARNING]

> > Be careful using `env(preferred-text-scale)` when `<meta name="text-scale" content="scale">` is set as this will result in text scaling being applied twice when combined with relative sizes, such as `em` and `rem`. For example, when the `<meta>` is set, a declaration such as `font-size: calc(2rem * env(preferred-text-scale))` will cause small font sizes to be made even smaller and large font sizes larger.

Reviewer: If the preferred text scale is 0.5, does  `calc(2rem * env(preferred-text-scale))` make the font smaller?

Reviewer 3: Yes. In that scenario, 2rem = 16px, so `calc(16 * 0.5)` becomes 8px. 😄

The `viewport-segment-*` variable names can be used to set your containers to fit neatly into the available segments of a multi-viewport-segment device such as a hinged or foldable device. The integers following the `viewport-segment-*` name indicate which segment of the multiple segments the environment variable is referencing.

> The `preferred-text-scale` variable can be used to size website text or other UI features proportionally to browser or OS-set font sizes. For example, you could set your body font size to be in proportion to a user-defined text scale like so:

Reviewer: ```suggestion
The `preferred-text-scale` variable can be used to size website text or other UI features proportionally to browser or OS-set font sizes, such as Chrome for Android's global OS "Font size" setting. For example, you could set the body font size to be a percentage based on the user-defined text scale:
```

definitely change "your" to "the". the rest is attempt to clarify what confused me when i first read this but was clarified by https://github.com/mdn/content/pull/43766/files#r3092107224

```css
body {
  font-size: calc(100% * env(preferred-text-scale));
}
```

For example, on Chrome for Android, the body font size will be set in proportion to the global OS "Font size" setting.

> 

Reviewer: ```suggestion
```

Reviewer: seemed random here, so suggest moving up (or removing)

Author: Yeah, we don't need this now, certainly not with your earlier edit.

> [!NOTE]
> Page sizes can also be set to be proportional to browser or OS-level font size by including [`<meta name="text-scale" content="scale">`](/en-US/docs/Web/HTML/Reference/Elements/meta/name/text-scale) inside the document `<head>`. The `<meta>` tag should be used in favor of the `env(preferred-text-scale)` when possible, as the `<meta>` tag is simpler to use and is supported across a wider range of platforms.
>

> > Avoid including both these features as this may result in text scaling being applied twice, with small font sizes being smaller and large font sizes being larger.

Reviewer 3: I guess it's more nuanced than this. While you don't want to combine it with font-relative units because then - yes - you will get the double scaling...

```css
font-size: calc(2rem * env(preferred-text-scale));
```

... it should be fine when it's multiplying an absolute length like `px`, because they aren't relative to the document's initial font size.

```css
font-size: calc(32px * env(preferred-text-scale));
```

Author: Thanks for the clarification, @JoshTumath!

I've updated this bit with a warning note:

> [!WARNING]
> Using both these features together should be done with caution as it may result in text scaling being applied twice. For example, when combining `em` or `rem` with `env(preferred-text-scale)` — `font-size: calc(2rem * env(preferred-text-scale))` — you will end up with small font sizes being smaller and large font sizes being larger. This won't be a problem when multiplying by an absolute length (for example, `px`), which isn't relative to the document's initial font size — for example `font-size: calc(32px * env(preferred-text-scale))`.

Does this make sense?

Reviewer 3: Yes that seems great!

> > Using both these features together should be done with caution as it may result in text scaling being applied twice. For example, when combining `em` or `rem` with `env(preferred-text-scale)` — `font-size: calc(2rem * env(preferred-text-scale))` — you will end up with small font sizes being smaller and large font sizes being larger. This won't be a problem when multiplying by an absolute length (for example, `px`), which isn't relative to the document's initial font size — for example `font-size: calc(32px * env(preferred-text-scale))`.

Reviewer: ```suggestion
> Be careful using `env(preferred-text-scale)` when `<meta name="text-scale" content="scale">`  is set as this will result in text scaling being applied twice when combined with relative sizes, such as  `em` and `rem`. For example, when the `<meta>` is set, a declaration such as `font-size: calc(2rem * env(preferred-text-scale))` will cause small font sizes to be made even smaller and large font sizes larger.
```
you can add the recommendation to use only the meta in that note
